edgedb-protocol/src/descriptors.rs

```diff
@@ -490,29 +546,96 @@ impl Descriptor {
             MultiRange(i) => &i.id,
             Enumeration(i) => &i.id,
             InputShape(i) => &i.id,
+            Object(i) => &i.id,
+            Compound(i) => &i.id,
             TypeAnnotation(i) => &i.id,
         }
     }
     pub fn decode(buf: &mut Input) -> Result<Descriptor, DecodeError> {
         <Descriptor as Decode>::decode(buf)
     }
+    pub fn normalize_to_base(
+        &self,
+        ctx: &query_arg::DescriptorContext,
+    ) -> Result<Descriptor, Error> {
+        let norm = match self {
+            Descriptor::Scalar(d) if d.base_type_pos.is_some() => {
+                match ctx.get(d.base_type_pos.unwrap())? {
+                    Descriptor::Scalar(d) => {
+                        &Descriptor::BaseScalar(BaseScalarTypeDescriptor { id: d.id.clone() })
+                    }
+                    desc => desc,
+                }
+            }
+            Descriptor::Scalar(d) => {
+                if ctx.proto.is_at_least(2, 0) {
+                    &Descriptor::BaseScalar(BaseScalarTypeDescriptor { id: d.id.clone() })
+                } else {
+                    unreachable!("scalar dereference to a non-base type")
+                }
+            }
+            desc => desc,
+        };
+
+        Ok(norm.clone())
+    }
+}
+
+impl<T: Decode> Decode for Vec<T> {
+    fn decode(buf: &mut Input) -> Result<Self, DecodeError> {
+        ensure!(buf.remaining() >= 2, errors::Underflow);
+        let element_count = buf.get_u16();
+        let mut elements = Vec::with_capacity(element_count as usize);
+        for _ in 0..element_count {
+            elements.push(T::decode(buf)?);
+        }
+        Ok(elements)
+    }
+}
+
+impl Decode for Option<u32> {
+    fn decode(buf: &mut Input) -> Result<Self, DecodeError> {
+        ensure!(buf.remaining() >= 4, errors::Underflow);
+
+        let val = match buf.get_i32() {
+            -1 => None,
+            n if n > 0 => Some(n as u32),
+            _ => errors::InvalidOptionU32.fail()?,
+        };
+
+        Ok(val)
+    }
+}
+
+impl Decode for TypePos {
+    fn decode(buf: &mut Input) -> Result<Self, DecodeError> {
+        ensure!(buf.remaining() >= 2, errors::Underflow);
+        Ok(Self(buf.get_u16()))
+    }
 }
 
 impl Decode for Descriptor {
     fn decode(buf: &mut Input) -> Result<Self, DecodeError> {
         use Descriptor as D;
+        if buf.proto().is_at_least(2, 0) {
```

Comment:
Could we create constants and/or methods for these? It would be useful for quickly searching for all parts of the decoding logic that are version-dependent:

ie:

```
fn is_modern(&self) -> bool;
```

or 

```
fn is_at_least_2(&self) -> bool;
```

Alternatively, we could use a constant for the versions:

```
const VERSION_MODERN: ProtocolVersion = ProtocolVersion::new(2, 0);
```

Ideally we shouldn't be using naked ints for discoverability.

Reply:
Added `is_2()`, because we have a similar `is_1()` already.